File: src/main/java/com/paragon/client/ui/configuration/old/impl/setting/SettingComponent.java
package com.paragon.client.ui.configuration.old.impl.setting;

import com.paragon.api.setting.Bind;
import com.paragon.api.setting.Setting;
import com.paragon.api.util.render.ITextRenderer;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.client.systems.module.impl.client.ClickGUI;
import com.paragon.client.systems.module.impl.client.Colours;
import com.paragon.client.ui.configuration.old.OldPanelGUI;
import com.paragon.client.ui.configuration.old.impl.module.ModuleButton;
import com.paragon.client.ui.util.animation.Animation;
import net.minecraft.client.Minecraft;
import org.lwjgl.opengl.GL11;

import java.awt.*;
import java.util.ArrayList;

/**
 * @author Wolfsurge
 */
public class SettingComponent<T> implements ITextRenderer {

    private final ModuleButton moduleButton;
    private final Setting<T> setting;
    private final float height;
    private final ArrayList<SettingComponent<?>> settingComponents = new ArrayList<>();
    public Animation animation;
    private float offset;

    public SettingComponent(ModuleButton moduleButton, Setting<T> setting, float offset, float height) {
        this.moduleButton = moduleButton;
        this.setting = setting;
        this.offset = offset;
        this.height = height;

        float settingOffset = getOffset();

        if (!setting.getSubsettings().isEmpty()) {
            for (Setting<?> setting1 : setting.getSubsettings()) {
                if (setting1.getValue() instanceof Boolean) {
                    getSettingComponents().add(new BooleanComponent(getModuleButton(), (Setting<Boolean>) setting1, settingOffset, height));
                    settingOffset += height;
                } else if (setting1.getValue() instanceof Bind) {
                    getSettingComponents().add(new KeybindComponent(getModuleButton(), (Setting<Bind>) setting1, settingOffset, height));
                    settingOffset += height;
                } else if (setting1.getValue() instanceof Number) {
                    getSettingComponents().add(new SliderComponent(getModuleButton(), (Setting<Number>) setting1, settingOffset, height));
                    settingOffset += height;
                } else if (setting1.getValue() instanceof Enum<?>) {
                    getSettingComponents().add(new ModeComponent(getModuleButton(), (Setting<Enum<?>>) setting1, settingOffset, height));
                    settingOffset += height;
                } else if (setting1.getValue() instanceof Color) {
                    getSettingComponents().add(new ColourComponent(getModuleButton(), (Setting<Color>) setting1, settingOffset, height));
                    settingOffset += height;
                }
            }
        }

        animation = new Animation(ClickGUI.getAnimationSpeed()::getValue, false, ClickGUI.getEasing()::getValue);
    }

    public void renderSetting(int mouseX, int mouseY) {
        if (!getSettingComponents().isEmpty() && hasVisibleSubsettings() && !(this instanceof ModeComponent)) {
            GL11.glPushMatrix();
            GL11.glScalef(0.5f, 0.5f, 0.5f);
            Minecraft.getMinecraft().fontRenderer.drawStringWithShadow("...", (getModuleButton().getPanel().getX() + getModuleButton().getPanel().getWidth() - 6.5f) * 2, (getModuleButton().getOffset() + getOffset() + 4f) * 2, -1);
            GL11.glPopMatrix();
        }

        if (animation.getAnimationFactor() > 0) {
            getSettingComponents().forEach(settingComponent -> {
                if (settingComponent.getSetting().isVisible()) {
                    settingComponent.renderSetting(mouseX, mouseY);
                }
            });

            for (SettingComponent<?> settingComponent : settingComponents) {
                if (settingComponent.getSetting().isVisible()) {
                    RenderUtil.drawRect(getModuleButton().getPanel().getX(), getModuleButton().getOffset() + settingComponent.getOffset(), 2, settingComponent instanceof ColourComponent ? getHeight() : settingComponent.getHeight(), Colours.mainColour.getValue().getRGB());
                }
            }
        }

        if (isMouseOver(mouseX, mouseY) && !(this instanceof ColourComponent)) {
            OldPanelGUI.tooltip = setting.getDescription();
        }

        // If it's a colour component, we only want to render if the mouse is over the actual button, not the pickers
        else if (this instanceof ColourComponent) {
            if (OldPanelGUI.isInside(getModuleButton().getPanel().getX(), getModuleButton().getOffset() + getOffset(), getModuleButton().getPanel().getX() + getModuleButton().getPanel().getWidth(), getModuleButton().getOffset() + getOffset() + 13, mouseX, mouseY)) {
                OldPanelGUI.tooltip = setting.getDescription();
            }
        }
    }

    public boolean hasVisibleSubsettings() {
        for (SettingComponent<?> settingComponent : getSettingComponents()) {
            if (settingComponent.getSetting().isVisible()) {
                return true;
            }
        }

        return false;
    }

    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if (mouseButton == 1) {
            if (isMouseOver(mouseX, mouseY)) {
                animation.setState(!isExpanded());
            }
        }
    }

    public void mouseReleased(int mouseX, int mouseY, int mouseButton) {

    }

    public void keyTyped(char typedChar, int keyCode) {
        for (SettingComponent<?> settingComponent : getSettingComponents()) {
            if (settingComponent.getSetting().isVisible()) {
                settingComponent.keyTyped(typedChar, keyCode);
            }
        }
    }

    public boolean isMouseOver(int mouseX, int mouseY) {
        return OldPanelGUI.isInside(getModuleButton().getPanel().getX(), getModuleButton().getOffset() + getOffset(), getModuleButton().getPanel().getX() + getModuleButton().getPanel().getWidth(), getModuleButton().getOffset() + getOffset() + getHeight(), mouseX, mouseY);
    }

    /**
     * Gets the module button
     *
     * @return The module button
     */
    public ModuleButton getModuleButton() {
        return moduleButton;
    }

    /**
     * Gets the setting
     *
     * @return The setting
     */
    public Setting<T> getSetting() {
        return setting;
    }

    /**
     * Gets the offset
     *
     * @return The offset
     */
    public float getOffset() {
        return offset;
    }

    /**
     * Sets the offset
     *
     * @param newOffset The new offset
     */
    public void setOffset(float newOffset) {
        this.offset = newOffset;
    }

    /**
     * Gets whether the setting component is expanded or not
     *
     * @return Whether the setting component is expanded or not
     */
    public boolean isExpanded() {
        return animation.getAnimationFactor() > 0;
    }

    /**
     * Gets the list of setting components
     *
     * @return The setting components
     */
    public ArrayList<SettingComponent<?>> getSettingComponents() {
        return settingComponents;
    }

    /**
     * Gets the height of the component (without subsettings)
     *
     * @return The height of the component
     */
    public float getHeight() {
        return height;
    }

    /**
     * Gets the absolute height of the module
     *
     * @return The absolute height of the module
     */
    public float getAbsoluteHeight() {
        // return getHeight() + (expanded ? subsettingHeight : 0);
        float subsettingHeight = 0;

        for (SettingComponent settingComponent : getSettingComponents()) {
            subsettingHeight += settingComponent.getHeight();
        }

        return isExpanded() ? getHeight() + subsettingHeight : getHeight();
    }
}

File: src/main/kotlin/com/paragon/client/ui/configuration/old/impl/setting/SettingComponent.kt
package com.paragon.client.ui.configuration.old.impl.setting

import com.paragon.api.setting.Bind
import com.paragon.api.setting.Setting
import com.paragon.api.util.render.RenderUtil.drawRect
import com.paragon.client.systems.module.impl.client.ClickGUI.animationSpeed
import com.paragon.client.systems.module.impl.client.ClickGUI.easing
import com.paragon.client.systems.module.impl.client.Colours
import com.paragon.client.ui.configuration.old.OldPanelGUI
import com.paragon.client.ui.configuration.old.OldPanelGUI.Companion.isInside
import com.paragon.client.ui.configuration.old.impl.module.ModuleButton
import me.surge.animation.Animation
import net.minecraft.client.Minecraft
import org.lwjgl.opengl.GL11
import java.awt.Color
import java.util.function.Consumer

/**
 * @author Wolfsurge
 */
open class SettingComponent<T>(val moduleButton: ModuleButton, val setting: Setting<T>, var offset: Float, val height: Float) {

    /**
     * Gets the list of setting components
     *
     * @return The setting components
     */
    val settingComponents = ArrayList<SettingComponent<*>>()
    var animation: Animation

    /**
     * Gets the offset
     *
     * @return The offset
     */

    init {
        var settingOffset = offset
        if (!setting.subsettings.isEmpty()) {
            for (setting1 in setting.subsettings) {
                if (setting1.value is Boolean) {
                    settingComponents.add(BooleanComponent(moduleButton, setting1 as Setting<Boolean>, settingOffset, height))
                    settingOffset += height
                }
                else if (setting1.value is Bind) {
                    settingComponents.add(KeybindComponent(moduleButton, setting1 as Setting<Bind?>, settingOffset, height))
                    settingOffset += height
                }
                else if (setting1.value is Number) {
                    settingComponents.add(SliderComponent(moduleButton, setting1 as Setting<Number?>, settingOffset, height))
                    settingOffset += height
                }
                else if (setting1.value is Enum<*>) {
                    settingComponents.add(ModeComponent(moduleButton, setting1 as Setting<Enum<*>>, settingOffset, height))
                    settingOffset += height
                }
                else if (setting1.value is Color) {
                    settingComponents.add(ColourComponent(moduleButton, setting1 as Setting<Color>, settingOffset, height))
                    settingOffset += height
                }
            }
        }
        animation = Animation(animationSpeed::value, false, easing::value)
    }

    open fun renderSetting(mouseX: Int, mouseY: Int) {
        if (!settingComponents.isEmpty() && hasVisibleSubsettings() && this !is ModeComponent) {
            GL11.glPushMatrix()
            GL11.glScalef(0.5f, 0.5f, 0.5f)
            Minecraft.getMinecraft().fontRenderer.drawStringWithShadow("...", (moduleButton.panel.x + moduleButton.panel.width - 6.5f) * 2, (moduleButton.offset + offset + 4f) * 2, -1)
            GL11.glPopMatrix()
        }
        if (animation.getAnimationFactor() > 0) {
            settingComponents.forEach(Consumer { settingComponent: SettingComponent<*> ->
                if (settingComponent.setting.isVisible()) {
                    settingComponent.renderSetting(mouseX, mouseY)
                }
            })
            for (settingComponent in settingComponents) {
                if (settingComponent.setting.isVisible()) {
                    drawRect(moduleButton.panel.x, moduleButton.offset + settingComponent.offset, 2f, if (settingComponent is ColourComponent) height else settingComponent.height, Colours.mainColour.value.rgb)
                }
            }
        }
        if (isMouseOver(mouseX, mouseY) && this !is ColourComponent) {
            OldPanelGUI.tooltip = setting.description
        }
        else if (this is ColourComponent) {
            if (isInside(moduleButton.panel.x, moduleButton.offset + offset, moduleButton.panel.x + moduleButton.panel.width, moduleButton.offset + offset + 13, mouseX, mouseY)) {
                OldPanelGUI.tooltip = setting.description
            }
        }
    }

    fun hasVisibleSubsettings(): Boolean {
        for (settingComponent in settingComponents) {
            if (settingComponent.setting.isVisible()) {
                return true
            }
        }
        return false
    }

    open fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        if (mouseButton == 1) {
            if (isMouseOver(mouseX, mouseY)) {
                animation.state = !isExpanded
            }
        }
    }

    open fun mouseReleased(mouseX: Int, mouseY: Int, mouseButton: Int) {}
    open fun keyTyped(typedChar: Char, keyCode: Int) {
        for (settingComponent in settingComponents) {
            if (settingComponent.setting.isVisible()) {
                settingComponent.keyTyped(typedChar, keyCode)
            }
        }
    }

    fun isMouseOver(mouseX: Int, mouseY: Int): Boolean {
        return isInside(moduleButton.panel.x, moduleButton.offset + offset, moduleButton.panel.x + moduleButton.panel.width, moduleButton.offset + offset + height, mouseX, mouseY)
    }


    open val isExpanded: Boolean
        get() = animation.getAnimationFactor() > 0// return getHeight() + (expanded ? subsettingHeight : 0);


    open val absoluteHeight: Float
        get() {
            var subsettingHeight = 0f

            for (settingComponent in settingComponents) {
                subsettingHeight += settingComponent.height
            }

            return if (isExpanded) height + subsettingHeight else height
        }
}